Cutting, restructuring, more explanations (#1)
kkohbrok/mls-associated-parties: draft-kohbrok-mls-associated-parties.md

} AssociatedParties
~~~

## Associated parties as external senders

Associated parties can act as external senders and thus send external proposals
as specified in Section 12.1.8 of {{!RFC9420}}. An AP that sends an external
proposal MUST use the `external` SenderType with an index that is the number of
entries in the group's `external_senders` (0 if no such extension is present)
plus its own index in the group's AssociatedParties extension. For external
sender indices, the AssociatedParties extension thus extends the
`external_senders` extension.

Recipients of a message sent by an AP MUST verify the message's signature

> against the `signature_key` in the sender's AssociatedPartyEntry.

Reviewer: I don't see that APs as external senders is strictly required. You could have the same entity in both the AP list and the external senders and everything would still work fine.

Author: Fair point. But then I'd want to mandate that APs are in the external senders list, because we want the HPKE public keys that we encrypt secrets to to be authenticated somehow.

Now that I think about it, having them in the external senders is in fact a much nicer approach. I'll change it.

struct {
  u32 removed_party_index;
} RemoveAssociatedParty

struct {
  AssociatedPartyEntry updated_party;
} UpdateAssociatedParty
~~~

> When a group member commits one or more of theses proposals, the

Reviewer: ```suggestion
When a group member commits one or more of these proposals, the
```

Both associated party and all other group members finally derive the
`ap_commit_secret_id` and the `ap_commit_secret`. They compare the
`ap_commit_secret_id` to the one they received along with the ciphertext to
ensure that everyone got the same base secret. The `ap_commit_secret` is used to
compute the new epoch of the associated party key schedule as described in
{{associated-party-key-schedule}}.

Open Question: Do we want a completely new message format here? It would
probably be easier to extend the `ContentType` enum, which would allow this
document to re-use the PublicMessage semantics. The content should then consist
of `kem_output`, `ciphertext and `ap_commit_secret_id`.

# Associated party key schedule

The secrets exported from the MLS key schedule and sent to the APs as described
in {{exporting-the-ap-secrets}} drive the AP key schedule, a simple key schedule
that mimics the MLS key schedule. In each epoch, the newly added key material is
combined with the key material from the previous epoch to derive the `ap_init`

> for the next epoch and the `ap_exporter`. The `ap_exporter` keys a puncturable

Reviewer: Verb missing here:
"The `ap_exporter` keys [???] a puncturable"

Author: Heh. "keys" was meant as a verb, but it's probably a bit unusual. I'll change it to something that makes the sentence easier to parse.

pseudorandom function (PPRF), which the application or other MLS extensions can
use to derive further secrets.

~~~ aasvg
                ap_init_secret_[n-1]
                        |
                        |
                        V
ap_commit_secret --> KDF.Extract
                        |
                        |
                        V
                ExpandWithLabel(., "ap_epoch", GroupContext_[n], KDF.Nh)
                        |
                        |
                        V
                  ap_epoch_secret
                        |
                        |
                        +--> DeriveSecret(., "ap_exporter")
                        |    = ap_exporter_secret
                        |
                        V
                  DeriveSecret(., "init")
                        |
                        |
                        V
                ap_init_secret_[n]
~~~
{: title="The Associated Party Key Schedule" #ap-key-schedule }

As the `ap_commit_secret` is distinct for each AP, each AP has its own key
schedule. Key material is thus shared between the group and each AP individually

> s.t. APs don't share key material with one-another.

Reviewer: ```suggestion
such that APs don't share key material with one-another.
```
